specifyweb/specify/api.py

```diff
@@ -642,39 +639,68 @@ def handle_to_many(collection, agent, obj, data: Dict[str, Any]) -> None:
-        ids = [] # Ids not in this list will be deleted at the end.
+        ids = [] # Ids not in this list will be deleted (if dependent) or removed from obj (if independent) at the end.
+        
+        ids_to_fetch = []
+        cached_objs = dict()
+        fk_model = None
+        if not is_dependent: 
+            for rel_data in val: 
+                if not isinstance(rel_data, str): continue
+                fk_model, fk_id = strict_uri_to_model(rel_data, rel_model.__name__)
+                ids_to_fetch.append(fk_id)
+
+            if fk_model is not None: 
+                cached_objs = {item.id: obj_to_data(item) for item in get_model(fk_model).objects.filter(id__in=ids_to_fetch)}
+
         for rel_data in val:
+            if isinstance(rel_data, str): 
+                assert not is_dependent, "expected object for dependent field %s in %s: %s" % (field_name, obj, rel_data)
+                fk_model, fk_id = strict_uri_to_model(rel_data, rel_model.__name__)
+                rel_data = cached_objs[fk_id]
+                if rel_data[field.field.name] == uri_for_model(obj.__class__, obj.id): 
+                    ids.append(rel_data["id"])
+                    continue
+
             rel_data[field.field.name] = obj
-            if 'id' in rel_data:
-                # Update an existing related object.
-                rel_obj = update_obj(collection, agent,
-                                     rel_model, rel_data['id'],
-                                     rel_data['version'], rel_data,
-                                     parent_obj=obj)
-            else:
-                # Create a new related object.
-                rel_obj = create_obj(collection, agent, rel_model, rel_data, parent_obj=obj)
+
+            rel_obj = update_or_create_resource(collection, agent, rel_model, rel_data, parent_obj=obj if is_dependent else None)
+
+            if not is_dependent and not (isinstance(obj, models.Recordset) and field_name == 'recordsetitems'):
+                getattr(obj, field_name).add(rel_obj)
             ids.append(rel_obj.id) # Record the id as one to keep.
 
         # Delete related objects not in the ids list.
         # TODO: Check versions for optimistic locking.
-        to_delete = getattr(obj, field_name).exclude(id__in=ids)
-        for rel_obj in to_delete:
-            check_table_permissions(collection, agent, rel_obj, "delete")
-            auditlog.remove(rel_obj, agent, obj)
-        to_delete.delete()
+        to_remove = getattr(obj, field_name).exclude(id__in=ids).select_for_update()
+        if is_dependent or (isinstance(obj, models.Recordset) and field_name == 'recordsetitems'):
+            for rel_obj in to_remove:
+                check_table_permissions(collection, agent, rel_obj, "delete")
+                auditlog.remove(rel_obj, agent, obj)
+        
+            to_remove.delete()
+        else: 
+            getattr(obj, field_name).remove(*list(to_remove))
```

Comment:
1. check permissions on related, you aren't.
2. do you really need to use the remove/add stuff? I don't like too much divergence from dependent code.